src/sinks/gcs_common/service.rs

```diff
@@ -0,0 +1,121 @@
+use crate::buffers::Ackable;
+use crate::event::{EventFinalizers, EventStatus, Finalizable};
+use crate::http::{HttpClient, HttpError};
+use crate::sinks::gcp::GcpCredentials;
+use bytes::Bytes;
+use futures::future::BoxFuture;
+use http::{
+    header::{HeaderName, HeaderValue},
+    Request, Uri,
+};
+use hyper::Body;
+use std::task::Poll;
+use tower::Service;
+
+#[derive(Debug, Clone)]
+pub struct GcsService {
+    client: HttpClient,
+    base_url: String,
+    creds: Option<GcpCredentials>,
+}
+
+impl GcsService {
+    pub const fn new(
+        client: HttpClient,
+        base_url: String,
+        creds: Option<GcpCredentials>,
+    ) -> GcsService {
+        GcsService {
+            client,
+            base_url,
+            creds,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct GcsRequest {
+    pub body: Bytes,
+    pub key: String,
+    pub settings: GcsRequestSettings,
+    pub finalizers: EventFinalizers,
+}
+
+impl Ackable for GcsRequest {
+    fn ack_size(&self) -> usize {
+        self.body.len()
+    }
+}
+
+impl Finalizable for GcsRequest {
+    fn take_finalizers(&mut self) -> EventFinalizers {
+        std::mem::take(&mut self.finalizers)
+    }
+}
+
+// Settings required to produce a request that do not change per
+// request. All possible values are pre-computed for direct use in
+// producing a request.
+#[derive(Clone, Debug)]
+pub struct GcsRequestSettings {
+    pub acl: Option<HeaderValue>,
+    pub content_type: HeaderValue,
+    pub content_encoding: Option<HeaderValue>,
+    pub storage_class: HeaderValue,
+    pub metadata: Vec<(HeaderName, HeaderValue)>,
+}
+
+#[derive(Debug)]
+pub struct GcsResponse {
+    pub inner: http::Response<Body>,
+}
+
+impl AsRef<EventStatus> for GcsResponse {
+    fn as_ref(&self) -> &EventStatus {
+        &EventStatus::Delivered
+    }
+}
```

Comment:
Once you merge in `master` and resolve the conflicts, the need for this specific impl will go away, but you'll need to impl `DriverReponse` instead, which is used by `Driver`.

Reply:
Updated.